src/hotspot/share/opto/vmaskloop.cpp

```diff
@@ -0,0 +1,975 @@
+  // Transform the loop and set flags
+  transform_loop(t_vmask);
+  cl->mark_loop_vectorized();
+  cl->mark_vector_masked();
+  _phase->C->set_max_vector_size(MaxVectorSize);
+  trace_msg(nullptr, "Loop is vector masked");
+}
+
+// ----------------------------------- Init -----------------------------------
+void VectorMaskedLoop::init(IdealLoopTree* lpt) {
+  // Set current loop info
+  _lpt = lpt;
+  _cl = lpt->_head->as_CountedLoop();
+  _cle = _cl->loopexit();
+  _iv = _cle->phi();
+
+  // Reset data structures
+  _core_set.clear();
+  _body_set.clear();
+  _body_nodes.clear();
+  _rpo_idx.clear();
+  _elem_bt.clear();
+  _stmts.clear();
+  _swptrs.clear();
+  _size_stats.clear();
+}
+
+// ------------------- Loop vectorizable analysis functions -------------------
+// Collect loop nodes into an array with reverse postorder for convenience of
+// future traversal. Do early bail out if unsupported node is found.
+bool VectorMaskedLoop::collect_loop_nodes() {
+  // Collect 7 (see EMPTY_LOOP_SIZE) core nodes of the loop
+  _lpt->collect_loop_core_nodes(_phase, _core_set);
+
+  // Push loop nodes into a node set for fast membership check, also create a
+  // temporary index map for RPO visit
+  int node_cnt = _lpt->_body.size();
+  for (int i = 0; i < node_cnt; i++) {
+    Node* n = _lpt->_body.at(i);
+    if (n->is_LoadStore() || n->is_RangeCheck() || n->is_Call()) {
+      trace_msg(n, "Found unsupported node in the loop");
+      return false;
+    }
+    _body_set.push(n);
+    set_rpo_idx(n, i);
+  }
+
+  // Visit all loop nodes from the head to create reverse postorder
+  VectorSet visited(_arena);
+  VectorSet post_visited(_arena);
+  GrowableArray<Node*> stack(_arena, node_cnt, 0, nullptr);
+  stack.push(_cl);
+  int idx = node_cnt - 1;
+  while (stack.length() > 0) {
+    Node* n = stack.top();
+    if (!visited.test(rpo_idx(n))) {
+      // Forward arc in graph
+      visited.set(rpo_idx(n));
+    } else if (!post_visited.test(rpo_idx(n))) {
+      // Cross or backward arc in graph
+      if (!is_memory_phi(n)) {
+        // Push all users in loop for non-mem-phi nodes
+        for (DUIterator_Fast imax, i = n->fast_outs(imax); i < imax; i++) {
+          Node* use = n->fast_out(i);
+          if (in_body(use) && !visited.test(rpo_idx(use))) {
+            stack.push(use);
+          }
+        }
+      }
+      if (n == stack.top()) {
+        // Node is still at the top - no additional use is pushed, visit it.
+        // Also initialize node info at this time.
+        stack.pop();
+        assert(idx >= 0, "Is some node visited more than once?");
+        _body_nodes.at_put_grow(idx, n);
+        _elem_bt.at_put_grow(idx, T_ILLEGAL);
+        idx--;
+        post_visited.set(rpo_idx(n));
+      }
+    } else {
+      stack.pop();
+    }
+  }
+
+  // Bail out if loop has unreachable node while traversing from head
+  if (idx != -1) {
+    trace_msg(nullptr, "Loop has unreachable node while traversing from head");
+    return false;
```

Comment:
Can this ever happen? Or could you add an assert here?

Reply:
Yes, it happened before. I will try to find a case.